File: src/server/scripts/DC/MapExtension/dc_map_gps.cpp
/*
 * DC-MapExtension GPS System
 * Sends player coordinates for custom zones to client
 * Uses Lua script execution (not AIO) for compatibility
 */

#include "ScriptMgr.h"
#include "Player.h"
#include "Map.h"
#include "WorldSession.h"
#include <sstream>
#include <iomanip>

// Custom zone map IDs
constexpr uint32 AZSHARA_CRATER_MAP_ID = 37;  // Confirmed via .gps command
constexpr uint32 HYJAL_MAP_ID = 1;  // Kalimdor, but check zone ID

// Custom zone IDs (if using zone detection)
constexpr uint32 AZSHARA_CRATER_ZONE_ID = 268;
constexpr uint32 HYJAL_ZONE_ID = 616;

class dc_map_gps_worldscript : public WorldScript
{
public:
    dc_map_gps_worldscript() : WorldScript("dc_map_gps_worldscript") { }

    void OnUpdate(uint32 diff) override
    {
        _updateTimer += diff;
        
        // Send GPS updates every 2 seconds
        if (_updateTimer >= 2000)
        {
            _updateTimer = 0;
            BroadcastGPSUpdates();
        }
    }

private:
    uint32 _updateTimer = 0;

    void BroadcastGPSUpdates()
    {
        // Iterate all players online
        SessionMap const& sessions = sWorld->GetAllSessions();
        for (SessionMap::const_iterator itr = sessions.begin(); itr != sessions.end(); ++itr)
        {
            if (Player* player = itr->second->GetPlayer())
            {
                if (!player->IsInWorld())
                    continue;

                SendGPSToPlayer(player);
            }
        }
    }
    
    void SendGPSToPlayer(Player* player)
    {
        uint32 mapId = player->GetMapId();
        uint32 zoneId = player->GetZoneId();
        
        // Only send GPS for custom zones
        bool isCustomZone = (mapId == AZSHARA_CRATER_MAP_ID) || 
                           (mapId == HYJAL_MAP_ID && zoneId == HYJAL_ZONE_ID);
        
        if (!isCustomZone)
            return;

        // Get player position
        float x = player->GetPositionX();
        float y = player->GetPositionY();
        float z = player->GetPositionZ();
        
        // Normalize coordinates to 0-1 range for map display
        float nx = 0.0f, ny = 0.0f;
        
        if (mapId == AZSHARA_CRATER_MAP_ID)
        {
            // Azshara Crater bounds: X roughly -1000 to 500, Y roughly -500 to 1500
            const float minX = -1000.0f, maxX = 500.0f;
            const float minY = -500.0f, maxY = 1500.0f;
            nx = (x - minX) / (maxX - minX);
            ny = (y - minY) / (maxY - minY);
        }
        else if (mapId == HYJAL_MAP_ID && zoneId == HYJAL_ZONE_ID)
        {
            // Hyjal bounds (adjust as needed based on actual zone coordinates)
            const float minX = -5000.0f, maxX = -3000.0f;
            const float minY = -2000.0f, maxY = 0.0f;
            nx = (x - minX) / (maxX - minX);
            ny = (y - minY) / (maxY - minY);
        }
        
        // Clamp to 0-1 range
        nx = std::max(0.0f, std::min(1.0f, nx));
        ny = std::max(0.0f, std::min(1.0f, ny));
        
        // Build Lua script to update GPS data on client
        std::ostringstream lua;
        lua << std::fixed << std::setprecision(3);
        lua << "if DCMapExtension_UpdateGPS then "
            << "DCMapExtension_UpdateGPS(" 
            << mapId << "," 
            << zoneId << "," 
            << x << "," 
            << y << "," 
            << z << "," 
            << nx << "," 
            << ny 
            << ") end";
        
        // Send Lua script to client
        WorldSession* session = player->GetSession();
        if (session)
        {
            WorldPacket data(SMSG_NOTIFICATION, lua.str().length() + 1);
            data << lua.str();
            session->SendPacket(&data);
        }
    }
};

void AddSC_dc_map_gps()
{
    new dc_map_gps_worldscript();
}

File: Custom/Eluna scripts/DC_MapGPS.lua
--[[
    DC-MapExtension GPS Broadcaster
    Sends player position data to DC-MapExtension client addon via AIO
    Updates every 2 seconds for players in custom zones
]]

local AIO = AIO or require("AIO")

-- Don't run on client
if not AIO.IsServer() then
    return
end

-- Custom zone definitions
-- IMPORTANT: These bounds must match the actual zone coordinates
-- Calibrated from all 4 corner positions (.gps measurements):
-- Azshara Crater corners:
--   Top-Left:     X=-1037.96,  Y=-1599.53
--   Top-Right:    X=1599.42,   Y=-1598.06
--   Bottom-Right: X=1567.81,   Y=1599.48
--   Bottom-Left:  X=-1059.23,  Y=1599.38
-- Map dimensions: 2658 x 3198 units (X × Y)
local CUSTOM_ZONES = {
    [37] = { -- Azshara Crater
        name = "Azshara Crater",
        zoneId = 268,
        -- Full corner measurements - map covers entire measured area
        -- Corners: X range [-1059, 1599], Y range [-1599, 1599]
        minX = -1059.0,
        maxX = 1599.0,
        minY = -1599.0,
        maxY = 1599.0
    },
    [1] = { -- Hyjal (requires zone check)
        name = "Hyjal",
        zoneId = 616,
        minX = 3600.0,  -- UPDATED: Positive X coordinates (was -5000)
        maxX = 5600.0,  -- UPDATED: Positive X coordinates (was -3000)
        minY = -4800.0, -- UPDATED: More negative Y (was -2000)
        maxY = -2800.0  -- UPDATED: More negative Y (was 0)
    }
}

-- Normalize coordinates to 0-1 range
local function NormalizeCoords(mapId, zoneId, x, y)
    local zone = CUSTOM_ZONES[mapId]
    if not zone then
        return nil, nil
    end
    
    -- For map 1 (Kalimdor), verify zone ID
    if mapId == 1 and zoneId ~= zone.zoneId then
        return nil, nil
    end
    
    -- Map 37 (Azshara Crater): X axis is flipped, Y axis is normal
    -- X=0,Y=0 should appear at 60.2%, 50.0% (verified in-game)
    if mapId == 37 then
        local nx = (zone.maxX - x) / (zone.maxX - zone.minX)  -- Flip X
        local ny = (y - zone.minY) / (zone.maxY - zone.minY)  -- Normal Y
        
        -- Clamp to 0-1
        nx = math.max(0, math.min(1, nx))
        ny = math.max(0, math.min(1, ny))
        
        return nx, ny
    else
        -- Normal calculation for other maps
        local nx = (x - zone.minX) / (zone.maxX - zone.minX)
        local ny = (y - zone.minY) / (zone.maxY - zone.minY)
        
        -- Clamp to 0-1
        nx = math.max(0, math.min(1, nx))
        ny = math.max(0, math.min(1, ny))
        
        return nx, ny
    end
end

-- Send GPS data to a player
local function SendGPS(player)
    if not player or not player:IsInWorld() then
        return
    end
    
    local mapId = player:GetMapId()
    local zoneId = player:GetZoneId()
    
    -- Check if player is in a custom zone
    if not CUSTOM_ZONES[mapId] then
        return
    end
    
    -- For map 1 (Kalimdor), verify it's actually Hyjal zone
    if mapId == 1 and zoneId ~= 616 then
        return
    end
    
    -- Get position
    local x, y, z = player:GetLocation()
    
    -- Normalize coordinates
    local nx, ny = NormalizeCoords(mapId, zoneId, x, y)
    if not nx or not ny then
        return
    end
    
    -- Build JSON string
    local json = string.format(
        '{"mapId":%d,"zoneId":%d,"x":%.3f,"y":%.3f,"z":%.3f,"nx":%.3f,"ny":%.3f}',
        mapId, zoneId, x, y, z, nx, ny
    )
    
    -- Send via AIO
    AIO.Handle(player, "DCMapGPS", "Update", json)
end

-- Broadcast GPS to all players in custom zones
local function BroadcastGPS()
    local players = GetPlayersInWorld()
    for _, player in ipairs(players) do
        SendGPS(player)
    end
end

-- Register timed event (every 2 seconds)
local GPS_UPDATE_INTERVAL = 2000 -- milliseconds
CreateLuaEvent(BroadcastGPS, GPS_UPDATE_INTERVAL, 0)

print("[DC-MapGPS] Eluna GPS broadcaster loaded - updating every 2 seconds")
